lib/eventFactory.js

```diff
@@ -2,9 +2,108 @@
+    });
+}
+
+/**
+ * Create builds associated with this event
+ * For example, if startFrom ~commit, then create builds with jobs that have requires: ~commit
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {Number}   [config.prNum]             PR number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               (Optional for backwards compatibility)
+ * @param  {Number}   eventId                    Event id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+    let jobsToStart = [];
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    return pipeline.jobs.then((jobs) => {
+        let nextJobs;
+
+        // When startFrom is ~commit or jobName
+        if (startFrom !== '~pr') {
+            nextJobs = WorkflowParser.getNextJobs(pipeline.workflowGraph, { trigger: startFrom });
+            jobsToStart = jobs.filter(j => nextJobs.includes(j.name) && j.state === 'ENABLED');
+
+            return Promise.resolve([]);
+        }
+
+        // Get next jobs for when startFrom is ~pr
+        nextJobs = WorkflowParser.getNextJobs(pipeline.workflowGraph, {
+            trigger: startFrom,
+            prNum: config.prNum
+        });
+
+        // To start all existing and ENABLED PR-prNum jobs
+        const PRJobArray = jobs.filter(j => j.name.startsWith(`PR-${config.prNum}:`));
+        const enabledPRJobArray = PRJobArray.filter(j => j.state === 'ENABLED');
+
+        jobsToStart = enabledPRJobArray;
+
+        // Get all the missing PR- job names
+        const existingPRJobNames = PRJobArray.map(p => p.name);
+        const missingPRJobNames = nextJobs.filter(j => !existingPRJobNames.includes(j));
+        // Get the job name part, e.g. main from PR-1:main
+        const missingJobNames = missingPRJobNames.map(name => name.split(':')[1]);
+        const missingJobs = jobs.filter(j => missingJobNames.includes(j.name));
+
+        // Create missing PR jobs
+        return Promise.all(missingJobs.map(j =>
+            createJob(jobFactory, j, `PR-${config.prNum}:${j.name}`)
```

Comment:
Shouldn't just steal the job config from pipeline cuz inside the PR ppl may change their job config. Should do the same logic currently in the webhook to get he config from prRef

https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/webhooks/index.js#L63